src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java

```diff
@@ -14,30 +14,171 @@
+  private static final String YOUTUBE_VIDEO_URL_PATTERN =
+      "(?<=watch\\?v=|/videos/|embed\\/|youtu.be\\/|\\/v\\/|\\/e\\/|"
+      + "watch\\?v%3D|watch\\?feature=player_embedded&v=|%2Fvideos%2F|"
+      + "embed%\u200C\u200B2F|youtu.be%2F|%2Fv%2F)[^#\\&\\?\\n]*";
+
+  /* Name of input field used for lecture name in lecture selection page. */
+  private static final String NAME_INPUT = "name-input";
+  
+  /* Name of input field used for lecture video link in lecture selection page. */
+  private static final String LINK_INPUT = "link-input";
+
+  private static final String REDIRECT_URL = "/lecture-view.html";
+
+  /* Pattern used to create a matcher for a video ID. */
+  private static Pattern videoUrlGeneratedPattern;
+
+  private static DatastoreService datastore;
+
   @Override
   public void init() throws ServletException {
-    // TODO: Implement LectureServlet.
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    videoUrlGeneratedPattern = Pattern.compile(YOUTUBE_VIDEO_URL_PATTERN);
   }
 
+  // TODO: Check if URL is valid.
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    Optional<String> optionalVideoUrl = getParameter(request, LINK_INPUT);
+    Optional<Entity> existingEntity = checkUrlInDatabase(optionalVideoUrl);
+
+    if (existingEntity.isPresent()) {
+      response.sendRedirect(buildRedirectUrl(existingEntity.get()).get());
+      return;
+    }
+    Entity lectureEntity = createLectureEntity(request);
+    datastore.put(lectureEntity);
+    response.sendRedirect(buildRedirectUrl(lectureEntity).get());
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    List<Lecture> lectures = getLectures();
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(lectures));
+  }
+
+  /**
+   * Returns the Entity in database that has {@code url}, or
+   * {@code Optional.empty()} if one doesn't exist.
+   */
+  private Optional<Entity> checkUrlInDatabase(Optional<String> videoUrl) {
+    String url = videoUrl.isPresent() ? videoUrl.get() : "";
+    if(url.isEmpty()) {
+      return Optional.empty();
+    }
+
+    Query query = new Query(Lecture.ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+    Iterable<Entity> resultsIterable = results.asIterable();
+
+    for (Entity lecture : resultsIterable) {
+      if (lecture.getProperty(Lecture.PROP_VIDEO_URL).equals(url)) {
+        return Optional.of(lecture);
+      }
+    }
+    return Optional.empty();
+  }
+
+  @VisibleForTesting
+  /** Creates and returns {@code lectureEntity} using parameters found in {@code request}. */
+  // TODO: Send errors as a response if fields are empty.
+  // TODO: Move to LectureEntityGenerator class.
```

Comment:
nit: keep the name of your class consistent (either LectureEntityCreator or LectureEntityGenerator)